File: frame/src/main/java/com/dreampany/frame/ui/model/BaseItem.java
package com.dreampany.frame.ui.model;

import android.content.Context;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.view.View;

import com.google.common.base.Objects;

import java.io.Serializable;

import eu.davidea.flexibleadapter.FlexibleAdapter;
import eu.davidea.flexibleadapter.items.AbstractFlexibleItem;
import eu.davidea.flexibleadapter.items.IFilterable;
import eu.davidea.viewholders.FlexibleViewHolder;

/**
 * Created by Hawladar Roman on 30/4/18.
 * Dreampany
 * dreampanymail@gmail.com
 */
public abstract class BaseItem<T, VH extends BaseItem.ViewHolder> extends AbstractFlexibleItem<VH> implements IFilterable, Serializable {

    protected T item;
    @LayoutRes
    protected int layoutId;
    protected boolean success;

    protected BaseItem(T item, int layoutId) {
        this.item = item;
        this.layoutId = layoutId;
        this.success = true;
    }

    @Override
    public boolean equals(Object inObject) {
        if (this == inObject) return true;
        if (inObject == null || getClass() != inObject.getClass()) return false;
        BaseItem item = (BaseItem) inObject;
        return Objects.equal(item, item.item);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(item);
    }

    public void setItem(T item) {
        this.item = item;
    }

    public void setLayoutId(@LayoutRes int layoutId) {
        this.layoutId = layoutId;
    }

    public void setSuccess(boolean success) {
        this.success = success;
    }

    @LayoutRes
    @Override
    public int getLayoutRes() {
        return layoutId;
    }

    @NonNull
    public T getItem() {
        return item;
    }

    public boolean isSuccess() {
        return success;
    }


    public static abstract class ViewHolder extends FlexibleViewHolder {

        public ViewHolder(View view, FlexibleAdapter adapter) {
            super(view, adapter);
        }

        public Context getContext() {
            return itemView.getContext();
        }
    }
}

File: frame/src/main/kotlin/com/dreampany/frame/ui/model/BaseItem.kt
package com.dreampany.frame.ui.model

import android.content.Context
import android.view.View
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView
import com.dreampany.frame.data.model.Base
import com.dreampany.frame.util.DisplayUtil
import com.google.common.base.Objects
import eu.davidea.flexibleadapter.FlexibleAdapter
import eu.davidea.flexibleadapter.items.AbstractFlexibleItem
import eu.davidea.flexibleadapter.items.IFilterable
import eu.davidea.flexibleadapter.items.IFlexible
import eu.davidea.viewholders.FlexibleViewHolder
import java.io.Serializable

/**
 * Created by Roman-372 on 7/11/2019
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
abstract class BaseItem<T : Base, VH : BaseItem.ViewHolder, S : Serializable>(var item: T, @LayoutRes var layoutId: Int = 0) :
    AbstractFlexibleItem<VH>(), IFilterable<S>, Serializable {

    var success: Boolean = false
    var favorite: Boolean = false
    var notify: Boolean = false
    var alert: Boolean = false
    var time: Long = 0L

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val item = other as BaseItem<T, VH, S>
        return Objects.equal(this.item, item.item)
    }

    override fun hashCode(): Int {
        return Objects.hashCode(item)
    }

    override fun getLayoutRes(): Int {
        return layoutId
    }

    override fun bindViewHolder(
        adapter: FlexibleAdapter<IFlexible<RecyclerView.ViewHolder>>,
        holder: VH,
        position: Int,
        payloads: MutableList<Any>?
    ) {
        holder.bind(position, this)
    }

    abstract class ViewHolder(val view: View, adapter: FlexibleAdapter<*>) :
        FlexibleViewHolder(view, adapter) {

        open fun getContext(): Context {
            return view.context
        }

        open fun <T : Base> setTag(tag: T) {
            view.setTag(tag)
        }

        open fun <T : Base> getTag(): T? {
            return view.tag as T?
        }

        open fun getSpanHeight(spanCount: Int, itemOffset: Int): Int {
            return (DisplayUtil.getScreenWidthInPx(getContext()) / spanCount) - (DisplayUtil.dpToPixels(
                itemOffset.toFloat()
            ) * spanCount)
        }

        abstract fun <VH : ViewHolder, T : Base, S : Serializable, I : BaseItem<T, VH, S>> bind(
            position: Int,
            item: I
        )

    }
}